src/prof_recent.c

```diff
@@ -383,14 +388,18 @@ prof_recent_alloc(tsd_t *tsd, edata_t *edata, size_t size) {
 }
 
 ssize_t
-prof_recent_alloc_max_ctl_read() {
-	/* Don't bother to acquire the lock. */
-	return prof_recent_alloc_max_get_no_lock();
+prof_recent_alloc_max_ctl_read(tsd_t *tsd) {
+	malloc_mutex_lock(tsd_tsdn(tsd), &prof_dump_mtx);
+	/* Don't bother to acquire the prof_recent_alloc_mtx lock. */
+	ssize_t max = prof_recent_alloc_max_get_no_lock();
+	malloc_mutex_unlock(tsd_tsdn(tsd), &prof_dump_mtx);
+	return max;
 }
 
-ssize_t
-prof_recent_alloc_max_ctl_write(tsd_t *tsd, ssize_t max) {
+static ssize_t
+prof_recent_alloc_max_write_locked(tsd_t *tsd, ssize_t max) {
 	assert(max >= -1);
+	malloc_mutex_assert_owner(tsd_tsdn(tsd), &prof_dump_mtx);
```

Comment:
Move the recent_alloc lock out of this function as well? Usually when I see locked, I assume all locks are taken care of.

Reply:
Right. I was trying to make it similar to `prof_recent_alloc_max_write_locked()`, which must be put under the dump lock but must not be under the recent lock. Should I also change the `prof_recent_alloc_max_write_locked()` name?

Reply:
I'd say still keep this function _locked(), i.e. and move the recent_mtx to the call sites: 1) the mallctl one can simply have the locking around the call; 2) the dumping one should be refactored so that it's called before we drop the recent_mtx in the last iteration?

Reply:
You mean `prof_recent_alloc_max_write_locked()` should be entered with the recent lock already been locked? If so it would need to temporarily drop the lock and re-lock if it needs to delete nodes, and in such cases the lock would have to be taken twice, rather than only once in the current logic.

Reply:
Ah I see. Then keep the current logic and only avoid the _locked in the name is fine.